### PR TITLE
Intersection option for WebGLRenderer clipping method

### PR DESCRIPTION
**Pull request by PWhiddy**
Currently custom clipping planes remove anything in the _union_ of the planes' clip space, however it is sometimes desirable to clip off only the _intersection_ of their clipping. 
This adds an option for WebGLRenderer's global clipping to use this method
Here are screenshots showing clipping

The regular way 

geometry 11

And with renderer.clipIntersection = true;

geometry 10

Not sure if @tschw has more plans for clipping in threejs, but it would also be nice if this option could be available at the local (material) clipping level

### CODE REVIEW

**Comment by mrdoob** [comment]
@tschw what do you think?

**Comment by Wilt** [comment]
I like this idea.

**Comment by tschw** [comment]
Nice! And the implementation looks good too.

Just as a thought, could allow the setting on a per-material basis at some later point...

**Comment by WestLangley** [comment]
An example of this enhancement would be great, too ! : - )

**Comment by mrdoob** [comment]
> An example of this enhancement would be great, too ! : - )

Yes please!

**Comment by PWhiddy** [comment]
@mrdoob @tschw 
Just added an example and added a line in the docs.
As tschw said, it would be nice eventually to have this available on a per-material basis, but I am not sure how best to implement this. For now local clipping does not work when the setting is enabled.

**Comment by PWhiddy** [comment]
@mrdoob Does it look good now?

**Comment by Wilt** [comment]
I like the example you created :+1:

**Comment by tschw** [comment]
@PWhiddy 

Very nice!

> As tschw said, it would be nice eventually to have this available on a per-material basis, but I am not sure how best to implement this. For now local clipping does not work when the setting is enabled.

It's the same code at shader level. Both data paths already exist in the renderer, so should be fairly straightforward after reading into it. Just come back with specific questions, in case you want to give it a go and run into problems.

> For now local clipping does not work when the setting is enabled.

Odd... Why? I figure it would just combine clipping differently in that case!?

**Comment by Wilt** [comment]
PWhiddy:
> For now local clipping does not work when the setting is enabled.
tschw:
> Odd... Why? I figure it would just combine clipping differently in that case!?

@tsch probably because of this line:

https://github.com/PWhiddy/three.js/blob/6d30e236ff210b679dff60ce3f0736d03883c66f/src/renderers/shaders/ShaderChunk/clipping_planes_fragment.glsl#L3

**Comment by PWhiddy** [comment]
> I like the example that you created :+1: 
The checkbox doesn't do anything though...

@Wilt The example doesn't work at that link because the threejs build is not updated in my PR (the docs say not to do this).

It will look like this with built library

screenshot 5

As for the local clipping, I think the reason that the global always overrides the local may be because in [WebGLClipping ](https://github.com/PWhiddy/three.js/blob/clip2/src/renderers/webgl/WebGLClipping.js) I always set both clipIntersection and scope.clipIntersection with the renderers global setting. 

I was not sure if it made more sense to add an extra clipIntersection parameter to all materials or if in the long term all materials should have just one property with an object that would contain planes (or different geometries), and other options bundled together. 

Something maybe like:

`material.clipping = { planes: [ p1, p2 ], boxes: [ b1 ], respectGlobal: false, clipIntersecton: true };`

I wasn't sure if this was the direction that clipping in threejs was headed, so I went with the simple version that maintains current functionality, but is more limited with the new feature for now.

**Comment by Wilt** [comment]
@PWhiddy Duh, that was silly... I like it anyway :smile:

**Comment by tschw** [comment]
For the global vs. local part, the intended use is global clipping for viewport restriction (e.g. portal culling) and local clipping for modeling purposes. That's why there's no shadow clipping for global planes (what is not visible can still cast a shadow) and no automatic object culling for local planes.

I guess it's fine to just have a global union vs. intersection switch, but it would be preferable to specify the planes on the materials in the examples to match the design rationale. Ideally, though we'd have to touch the GLSL for that, intersection mode would only apply to local planes.

**Comment by PWhiddy** [comment]
@tschw @mrdoob 
I finally was able to make the clipping option available locally on materials rather than globally!
Also updated the example and the API docs. 
Tested the advanced clipping example and everything seems to be working.
Feel free to check for for yourselves, but should be ready to merge!

**Comment by tschw** [comment]
Interesting...

I find the interplay of global and local planes with this implementation somewhat questionable: The material setting now dictates how global clipping planes are to be interpreted. Does this make any sense? I fail to find a use case for these semantics - not even a far-fetched one, which does not necessarily mean there is none. 
Since this behavior is also undocumented, I get the feeling it's random weirdness, driven by the structure of the existing implementation. Then however, making things even weirder, the local setting *is* considered during FOV culling, at this point only for global planes, which is correct - too correct for being undefined / accidental behavior or an intentional limitation. So I'm left a little puzzled. Please correct me in case I'm missing something.

If there is a union vs. intersection switch on a per-material basis, I think it shouldn't affect the global planes and the global union vs. intersection setting. Basically what this means is that the vertex shader would have to have two independent loops checking for clipping - one for the global planes and one for the local planes. The vec4 array uniform can be shared, since global planes always come first, however, both code blocks would have to be conditional, switched off based on the presence of respective global or local planes.

The more limited alternative (implementation-wise simpler than this code, so I'm kind-of sorry to say) would be a global union vs. intersection switch that just applies to all the planes.

A global switch only applying to the local planes would even be nicer, in this case, changes to FOV culling (a likely hotspot for complex scenes) could be left out (since it only considers global planes) but again aforementioned vertex shader modifications would be required.

All these three variants would make more sense to me than what we see implemented here. The second one being the weakest, but easiest to implement from where we're at. I haven't quite decided on whether I like the first or the third better: The first is very complete but needs an extra material property, which BTW also needs serialization support - missing here, the third OTOH is more limited but should suffice to capture the most typical use cases reasonably well.

**Comment by PWhiddy** [comment]
Thanks for your detailed response! I did not intentionally mean for the per-material clipping mode to affect the global planes, although now I think I see your point that since local and global are currently combined before being given to the shader they would need to be separated. 

I think the second method you described was basically my first implementation from August (which doesn't appear to be visible here because my recent pushes must have overridden it) 
I could revert to this if it's acceptable for the library, or perhaps if you gave me a couple pointers (maybe describe the changes that would be needed) I could try to implement the first or third option (would just vertex need modification or fragment too?). 
As to which variant would be most practical, if you decide which you like best I'm willing to take a shot at it.
Also good catch on the serialization, I can update that as well.

**Comment by tschw** [comment]
> would just vertex need modification or fragment too?

Confused me: In fact, it's only in the fragment shader. Would be, if WebGL had clipping support like desktop GL has.

> if you gave me a couple pointers (maybe describe the changes that would be needed) I could try to implement the first or third option (would just vertex need modification or fragment too?

OK, after thinking about this a bit, it could actually end up a lot easier than it sounds described in my previous post:

I think it makes sense to have global clipping restricted to the union mode. Then there is no need to touch FOV culling (a likely hotspot) and also simplifies the implementation as described in the next paragraph.

Local clipping may either clip the union or the intersection of the planes, so shader-wise it comes down to just adding code that does the intersection part - without having it affect union clipping. Since the global planes are always the first ones in the array that goes into the uniform and aforementioned restriction of global planes to union mode, we can use a single define to control it all (just like in your previous implementation, trading the flag for an integer): 

```glsl
// DISCLAIMER: For illustration - keep the details like in the existing code.
#if NUM_CLIP_PLANES_UNION
for ( int i = 0; i < NUM_CLIP_PLANES_UNION; ++i )
    if ( dot( clipPlanes[ i ], coord ) < 0.0 ) discard;
#endif

#if NUM_CLIP_PLANES_UNION < NUM_CLIP_PLANES_TOTAL
for ( int i = NUM_CLIP_PLANES_UNION; i < NUM_CLIP_PLANES_TOTAL; ++i ) {
    if ( dot( clipPlanes[ i ], coord ) < 0.0 ) break;
    if ( i == NUM_CLIP_PLANES_TOTAL - 1 ) discard;
}
#endif
```

Now for clipping the union of all local planes (like now), `NUM_CLIP_PLANES_UNION = NUM_CLIP_PLANES_TOTAL`.

For clipping the intersection of all local planes, `NUM_CLIP_PLANES_UNION` would be the number of global clipping planes.

These values are around in the renderer code paths for updating - and I guess it should be fairly easy to get it in. Transform to view space would be in place without having to change anything. Only thing left to decide is whether to control the local mode globally or on a per-material basis.

**Comment by PWhiddy** [comment]
I like your idea, it is a very good solution! 
Here is a new implementation, following this idea. Global clipping always uses union mode, and the clipIntersection material property applies only to local clipping.
This works well because the shader clipping loop is essentially unchanged when intersection is not enabled, and when it is the clipping loop is simply split into two parts and they share the same uniform array.

How does it look?

**Comment by PWhiddy** [comment]
Also, you previously mentioned serialization support for the new material property. I added the clipIntersection property to the "copy" function, but does it also need to be added to toJSON? It seems quite a few properties are missing there

**Comment by mrdoob** [comment]
@tschw looks good?

File: src/renderers/webgl/WebGLClipping.js [hunk 1]
```diff
@@ -151,6 +154,7 @@ function WebGLClipping() {
 		}
 
 		scope.numPlanes = nPlanes;
+		// Add clipIntersection
```
Comment (tschw) [inline src/renderers/webgl/WebGLClipping.js]: Don't get this comment. Leftover?

**Review by tschw — APPROVED** [review]
Looks good.

**Comment by tschw** [comment]
> does it also need to be added to toJSON

I just double-checked. I though the planes were in the JSON already, but it seems they are currently not serialized, so I guess we also don't have to serialize the flag.

The reason for that, and also why I've been so insistent on making both modes able to coexist, is that we may want to lift the planes off the material to become nodes in the scene graph that clip their children at some point.

File: src/renderers/webgl/WebGLClipping.js (hunk 1 above)
Comment (PWhiddy) [inline src/renderers/webgl/WebGLClipping.js]: Yeah it was leftover. Just removed it

**Comment by PWhiddy** [comment]
Okay, that makes sense. 

Thank you for all the guidance tschw, it was very helpful and wouldn't have been possible otherwise!

**Comment by mrdoob** [comment]
Thanks guys!

**Pull request merged**